[WebGPU] Add CompositorIntegration object

litherum opened the pull request:
#### 98794f128842c07c91cf7048e3a9e574d701f56a
<pre>
[WebGPU] Add CompositorIntegration object
<a href="https://bugs.webkit.org/show_bug.cgi?id=250955">https://bugs.webkit.org/show_bug.cgi?id=250955</a>
rdar://104517402

Reviewed by Tadeu Zagallo.

Right now, the data flow that WebGPU uses to interact with the compositor doesn&apos;t match either the
WebGPU spec&apos;s API nor WebGPU.h. Ideally, we&apos;d align them, for conceptual simplicity and to make
future hacking on this system easier.

The way this is supposed to work is the Surface gets created with a reference to a compositor
integration object. In the simple case, that&apos;s a CAMetalLayer, but in WebKit we don&apos;t use CAMetalLayer.
Instead, this patch creates an analogue for CAMetalLayer that contains the functionality to
integrate WebGPU with WebKit&apos;s compositor. Having this object means we can (eventually) get rid of
the custom APIs added to WebGPU.framework which cause leaky abstractions.

Just like CAMetalLayer, this object, named CompositorIntegration, is the owner of multiple IOSurfaces
used for render destinations. It has an API to recreate its IOSurfaces, which is mean to be called
whenever their attributes need to be changed. (For WebGPU, the creation of a swap chain indicates that
the IOSurfaces&apos; attributes need to be changed.) CompositorIntegration also has an IPC message receiver,
like almost all of the WebGPU types, and has one exposed message which returns the current IOSurfaces.
This is meant to be called just after they get recreated, so GPUSwapChain can get access to the
IOSurfaces, so GPUSwapChain can set them as the appropriate layer&apos;s contents.

There is no WGPU analog of CompositorIntegration; just like CAMetalLayer, it exists outside of
WebGPU.framework.

Other than the creation and retention of IOSurfaces, there&apos;s nothing really special about this object;
it&apos;s set up pretty much identically to every other WebGPU object.

The use of CompositorIntegration isn&apos;t hooked up yet; I made a megapatch at
<a href="https://github.com/WebKit/WebKit/pull/8931/files">https://github.com/WebKit/WebKit/pull/8931/files</a> and I&apos;m trying to split it up into smaller, more
reviewable patches. This patch is just one piece of that larger patch. CompositorIntegration will be
hooked up in a subsequent patch.

* Source/WebCore/Modules/WebGPU/GPUCompositorIntegration.cpp: Added.
(WebCore::GPUCompositorIntegration::getRenderBuffers const):
* Source/WebCore/Modules/WebGPU/GPUCompositorIntegration.h: Added.
(WebCore::GPUCompositorIntegration::create):
(WebCore::GPUCompositorIntegration::backing):
(WebCore::GPUCompositorIntegration::backing const):
(WebCore::GPUCompositorIntegration::GPUCompositorIntegration):
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/CMakeLists.txt:
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUCompositorIntegrationImpl.cpp: Added.
(PAL::WebGPU::CompositorIntegrationImpl::CompositorIntegrationImpl):
(PAL::WebGPU::CompositorIntegrationImpl::getRenderBuffers):
(PAL::WebGPU::toCFNumber):
(PAL::WebGPU::CompositorIntegrationImpl::recreateIOSurfaces):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUCompositorIntegrationImpl.h: Added.
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUCompositorIntegration.h: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCompositorIntegration.cpp: Added.
(WebKit::RemoteCompositorIntegration::RemoteCompositorIntegration):
(WebKit::RemoteCompositorIntegration::stopListeningForIPC):
(WebKit::RemoteCompositorIntegration::getRenderBuffers):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCompositorIntegration.h: Added.
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCompositorIntegration.messages.in: Added.
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteCompositorIntegrationProxy.cpp: Added.
(WebKit::WebGPU::RemoteCompositorIntegrationProxy::RemoteCompositorIntegrationProxy):
(WebKit::WebGPU::RemoteCompositorIntegrationProxy::getRenderBuffers):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteCompositorIntegrationProxy.h: Added.

Canonical link: <a href="https://commits.webkit.org/259588@main">https://commits.webkit.org/259588@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/829f7b38317b9ff37685db80d3f90f68013ecaab

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105345 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14419 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38224 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114602 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/174783 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/109248 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15603 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5351 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97653 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/114504 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111104 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/12065 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/95044 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/39546 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/93931 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/26678 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/81216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/7743 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/28038 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7853 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/4629 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/13887 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/47582 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6612 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9642 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski reviewed (changes requested):
I think this patch adds unnecessary complexity to what is already a complex system. Instead I think we should go the other way, remove Surface or SwapChain and have a single Canvas object which would be the analogue to CAMetalLayer

This is the direction the specification has already headed with GPUCanvasContext.

litherum committed:
[WebGPU] Add CompositorIntegration object
https://bugs.webkit.org/show_bug.cgi?id=250955
rdar://104517402

Reviewed by Tadeu Zagallo.

Right now, the data flow that WebGPU uses to interact with the compositor doesn't match either the
WebGPU spec's API nor WebGPU.h. Ideally, we'd align them, for conceptual simplicity and to make
future hacking on this system easier.

The way this is supposed to work is the Surface gets created with a reference to a compositor
integration object. In the simple case, that's a CAMetalLayer, but in WebKit we don't use CAMetalLayer.
Instead, this patch creates an analogue for CAMetalLayer that contains the functionality to
integrate WebGPU with WebKit's compositor. Having this object means we can (eventually) get rid of
the custom APIs added to WebGPU.framework which cause leaky abstractions.

Just like CAMetalLayer, this object, named CompositorIntegration, is the owner of multiple IOSurfaces
used for render destinations. It has an API to recreate its IOSurfaces, which is mean to be called
whenever their attributes need to be changed. (For WebGPU, the creation of a swap chain indicates that
the IOSurfaces' attributes need to be changed.) CompositorIntegration also has an IPC message receiver,
like almost all of the WebGPU types, and has one exposed message which returns the current IOSurfaces.
This is meant to be called just after they get recreated, so GPUSwapChain can get access to the
IOSurfaces, so GPUSwapChain can set them as the appropriate layer's contents.

There is no WGPU analog of CompositorIntegration; just like CAMetalLayer, it exists outside of
WebGPU.framework.

Other than the creation and retention of IOSurfaces, there's nothing really special about this object;
it's set up pretty much identically to every other WebGPU object.

The use of CompositorIntegration isn't hooked up yet; I made a megapatch at
https://github.com/WebKit/WebKit/pull/8931/files and I'm trying to split it up into smaller, more
reviewable patches. This patch is just one piece of that larger patch. CompositorIntegration will be
hooked up in a subsequent patch.

* Source/WebCore/Modules/WebGPU/GPUCompositorIntegration.cpp: Added.
(WebCore::GPUCompositorIntegration::getRenderBuffers const):
* Source/WebCore/Modules/WebGPU/GPUCompositorIntegration.h: Added.
(WebCore::GPUCompositorIntegration::create):
(WebCore::GPUCompositorIntegration::backing):
(WebCore::GPUCompositorIntegration::backing const):
(WebCore::GPUCompositorIntegration::GPUCompositorIntegration):
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/CMakeLists.txt:
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUCompositorIntegrationImpl.cpp: Added.
(PAL::WebGPU::CompositorIntegrationImpl::CompositorIntegrationImpl):
(PAL::WebGPU::CompositorIntegrationImpl::getRenderBuffers):
(PAL::WebGPU::toCFNumber):
(PAL::WebGPU::CompositorIntegrationImpl::recreateIOSurfaces):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUCompositorIntegrationImpl.h: Added.
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUCompositorIntegration.h: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCompositorIntegration.cpp: Added.
(WebKit::RemoteCompositorIntegration::RemoteCompositorIntegration):
(WebKit::RemoteCompositorIntegration::stopListeningForIPC):
(WebKit::RemoteCompositorIntegration::getRenderBuffers):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCompositorIntegration.h: Added.
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCompositorIntegration.messages.in: Added.
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteCompositorIntegrationProxy.cpp: Added.
(WebKit::WebGPU::RemoteCompositorIntegrationProxy::RemoteCompositorIntegrationProxy):
(WebKit::WebGPU::RemoteCompositorIntegrationProxy::getRenderBuffers):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteCompositorIntegrationProxy.h: Added.

Canonical link: https://commits.webkit.org/259588@main
